docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,385 @@
+a result, fully qualified paths such as `instance::execution::Domain` may be
+used below, and not have corresponding rustdoc - these items exist in the crate
+source though!
+
+As much as is possible, `KillSwitch` tries to be self-contained; no members are
+public, and it tries to avoid leaking details of its state-keeping into public
+interfaces. Currently, `lucet-runtime` is heavily dependent on POSIX
+thread-directed signals to implement guest timeouts - for non-POSIX platforms
+alternate implementations may be plausible, but need careful consideration of
+the race conditions that can arise from other platform-specific functionality.
+
+`KillSwitch` fundamentally relies on two pieces of state for safe operation,
+which are encapsulated in a `KillState` held by the `lucet_runtime::Instance`
+it terminates:
+* `execution_domain`, a `Domain` that describes the kind of execution that is
+  currently happening in the `Instance`.
+  -  This is kept in a `Mutex` since in many cases it will need to be accessed
+     either by a `KillSwitch` or `KillState`, and for the duration either are
+     considering the domain, it must block other users.
+* `terminable`, an `AtomicBool` that indicates if the `Instance` may stop
+  executing.
+  - `terminable` is in some ways a subset of the information expressed by
+    `Domain`. It is true if and only if the instance is not obligated to a
+    specific exit mechanism yet, which could be determined by examing the
+    active `Domain` at points we check `terminable` instead. Even though this
+    is duplicative, it is necessary for a correct implementation when POSIX
+    signal handlers are involved, because it is extremely inadvisable to take
+    locks in a signal handler. While a `Mutex` is can be dangerous, we can
+    share an `AtomicBool` in signal-safety-constrained code with impunity! See
+    the section `Timeout while handling a guest fault` for more details on
+    working under this constraint.
+
+The astute observer may notice that `Lucet` contains both an instance `Domain`
+and an instance `State`. These discuss different aspects of the instance's
+lifecycle: `Domain` describes what the instance is doing right now, where
+`State` describes the state of the instance as visible outside `lucet_runtime`.
+This is fundamentally the reason why `State` does not and cannot describe a
+"making a hostcall" variant - this is not something `lucet_runtime` wants to
+expose, and at the moment we don't intend to make any commitments about being
+able to query instance-internal state like this.  Additionally, this is why
+`Domain::Terminated` only expresses that an instance has _stopped_ running, not
+if it successfully or unsuccessfully exited. From `Domain`'s perspective, the
+instance has stopped running, and that's all there is to it. For more
+information, the `Instance`'s `State` will describe how the instance stopped,
+and if that's by a return or a fault. TODO: there's even an argument that
```

Comment:
I notice this TODO.

Reply:
This is gone now. I actually think `Cancelled` vs `Terminated` is a rough spot in our model of termination, and wanted to say as much in this document, but wasn't sure how to phrase better than a TODO. @katie-martin-fastly understandably removed it in an edit and I'll just do a PR later this week to actually try the simplification.